### PR TITLE
One Elasticsearch index to rule them all

### PR DESCRIPTION
**Pull request by KatrinIhler**
(Is the joke getting old yet?)

With this PR, there will be only one Elasticsearch index supporting both the Admin UI and the External API. The old index implementations were removed, the new one replaced the AbstractSearchIndex class that already contained most of the relevant functionality. The index-rebuild module was merged with the elasticsearch-index module for simplification.

The endpoints to clear and rebuild the index were moved to /index. Both the Admin UI and External API endpoints related to that were removed. This might slightly violate the External API versioning, but there was no good way to keep this backwards-compatible, and you shouldn't trigger an index rebuild from an external application anyway.

The new index has a new identifier that Opencast uses to find it. When updating to OC 11, there will be three different ways to get this to work (can you tell that I couldn't decide which one was better?)

1. Just start an index rebuild. A simple way to go for smaller systems.
2. If an index rebuild will not be required for other reasons, you can use an included migration script to rename your admin UI index. There might be some things to consider about the cloning process used if you have a complex Elasticsearch setup, but for most people this should be straightforward.
3. Both too much work? If you don't care about consistency, you can use the new config option to just keep using your old admin UI index without renaming it.

Additionally, there's a second script to remove the old indices if you want to. I hope I made it sufficiently clear that that should be used with caution. 

I hope I added/updated all the relevant docs. I cleared the release notes for OC11, I think that was the right thing to do?

While I was at it, I did some very minor cleanup where unnecessary complex method parameters and unused classes were concerned. I also killed the EventIndex interface.

I added some configuration and split up the index name into name and identifier (one used for logging, one to find the actual index), because I felt they had different use cases. In general, I struggled with some indecisiveness where naming was concerned, so I hope the commits aren't too confusing. If you struggle understanding something, just ask.

PS: Forgot to add the best thing - the index rebuild should now be twice as fast ;)

### CODE REVIEW

File: docs/guides/admin/docs/releasenotes/filter-admin-ui-users.txt
Comment (lkiesow) [inline docs/guides/admin/docs/releasenotes/filter-admin-ui-users.txt]: Why does this delete any release notes?

File: docs/guides/admin/docs/upgrade.md
Comment (lkiesow) [inline docs/guides/admin/docs/upgrade.md]:
```suggestion
    curl -i -u <admin_user>:<password> -s -X POST \
```

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/delete-old-indices.sh]:
```suggestion
		curl -u "$USER:$PASSWORD" -X POST  "$SERVER/$DELETE_INDEX/_open" -w "\n\n" -i -s
```

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/delete-old-indices.sh]:
```suggestion
		RESPONSE=$(curl -u "$USER:$PASSWORD" -X DELETE  "$SERVER/$DELETE_INDEX" -i -s)
```

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/delete-old-indices.sh]:
```suggestion
set -eu
```

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/delete-old-indices.sh:38]: Maybe use `curl -f …`?

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/delete-old-indices.sh]:
```suggestion
curl -u "$USER:$PASSWORD" -X GET "$SERVER/_cat/indices?v" -w "\n"
```

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/delete-old-indices.sh]:
```suggestion
curl -u "$USER:$PASSWORD" -X GET "$SERVER/_cat/indices?v" -w "\n"
```

File: docs/upgrade/10_to_11/migrate-indices.sh
Comment (lkiesow) [inline docs/upgrade/10_to_11/migrate-indices.sh]:
```suggestion
set -eu
```

Plus all the `$USER:$PASSWORD` should be in quotes

**Review by lkiesow — DISMISSED** [review]
Faster index rebuilds and less duplication :tada:

File: docs/guides/admin/docs/releasenotes/filter-admin-ui-users.txt
Comment (KatrinIhler) [reply docs/guides/admin/docs/releasenotes/filter-admin-ui-users.txt]: I was already wondering whether that was the right way to go - do we want to keep these for older releases? If I remember correctly, the extra files are just for the release managers so they can easily create the release notes, so having them in there for multiple releases without distinction might be very confusing? On the other hand, these are extra files so merging forward is easier, if I remember correctly, so... Idk, I'm confused. These wouldn't be new features in OC11, so do we want to keep them in here or not? I'm fine with either.

**Comment by KatrinIhler** [comment]
@lkiesow Thanks for the review, I'm gonna do the necessary changes sometime next week! :)

**Comment by gregorydlogan** [comment]
> I hope I added/updated all the relevant docs. I cleared the release notes for OC11, I think that was the right thing to do?

This seems reasonable to me.  In theory this is the job of the RM, but we don't have one formally appointed yet, and I'm not sure it matters anyway :)

**Comment by KatrinIhler** [comment]
Fix for the database test failure is waiting for review at #2877. Will rebase once that's merged. :)

File: docs/upgrade/10_to_11/delete-old-indices.sh
Comment (KatrinIhler) [reply docs/upgrade/10_to_11/delete-old-indices.sh:38]: `-f` swallows the response body, and I thought that information might be helpful to some people. There is `--fail-with-body` now, but it's very new, my cURL for example can't do that yet...

**Comment by KatrinIhler** [comment]
@lkiesow I think I addressed all of your review findings now. :)

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md]: So, can I just do nothing and ignore these instructions completely?  My read of the previous sentence is that I need to either rebuild, or move/migrate my existing index.  The last sentence implies that I can do nothing, and suffer from reduced functionality in the external API (ie, its index is missing), but otherwise stuff will work as it previously did.

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md]: This step needs to be done *after* starting Opencast if you're rebuilding, right?

Perhaps
```
please refer to [older release notes](...)

Bold font: Note: Only one of steps 4, or 6 are required.

1. Stop your current Opencast install
...
4. Optionally [Migrate Elasticsearch index]... -> linking directly to the relevant part
5. Start Opencast
6. Optionally [Rebuild Elasticsearch index]... -> linking directly to the relevant part
```

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md:28]: Changes to step list means we can remove the Starting of Opencast I think.

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md]: IMO this should be moved down to the important notices section below.  Don't tell people there's a cleanup script before telling them why they shouldn't run it without testing things first :)

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md]: I know where the subindices live, but I would probably give a hint on how to do this.

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md]: This, taken with the `Do not start` line above should, imo, be combined into something like:

`- The migration process is an offline process that cannot be safely accomplished while Opencast is running.  Ensure Opencast is stopped on the admin node prior to starting this migration.`

File: docs/guides/admin/docs/upgrade.md
Comment (gregorydlogan) [inline docs/guides/admin/docs/upgrade.md:74]: How does one do this?  Same script as above?

**Review by gregorydlogan — CHANGES_REQUESTED** [review]
LGTM, minor doc nitpicks.  Also ran into this:
```
2021-09-29T15:16:19,499 | INFO  | (IndexEndpoint:112) - Clear the index
2021-09-29T15:16:29,976 | INFO  | (IndexEndpoint:158) - Starting to repopulate the index
2021-09-29T15:16:30,235 | ERROR | (AbstractEventEndpoint:2435) - The admin UI Search Index was not able to get the events list:
org.opencastproject.elasticsearch.api.SearchIndexException: Error querying event index
	at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:584) ~[!/:?]
	at org.opencastproject.adminui.endpoint.AbstractEventEndpoint.getEvents(AbstractEventEndpoint.java:2433) [!/:?]
	at jdk.internal.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
...
Caused by: org.opencastproject.elasticsearch.api.SearchIndexException: [opencast_event] ElasticsearchStatusException[Elasticsearch exception [type=index_not_found_exception, reason=no such index [opencast_event]]]
	at org.opencastproject.elasticsearch.impl.AbstractElasticsearchIndex.executeQuery(AbstractElasticsearchIndex.java:536) ~[!/:?]
	at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:576) ~[!/:?]
	... 115 more
Caused by: org.elasticsearch.ElasticsearchStatusException: Elasticsearch exception [type=index_not_found_exception, reason=no such index [opencast_event]]
	at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:187) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:1911) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:1888) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1645) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1602) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1572) ~[!/:?]
	at org.elasticsearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1088) ~[!/:?]
	at org.opencastproject.elasticsearch.impl.AbstractElasticsearchIndex.executeQuery(AbstractElasticsearchIndex.java:534) ~[!/:?]
	at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:576) ~[!/:?]
	... 115 more
	Suppressed: org.elasticsearch.client.ResponseException: method [POST], host [http://localhost:9200], URI [/opencast_event/_search?typed_keys=true&max_concurrent_shard_requests=5&ignore_unavailable=false&expand_wildcards=open&allow_no_indices=true&preference=_local&ignore_throttled=true&search_type=query_then_fetch&batched_reduce_size=512&ccs_minimize_roundtrips=true], status line [HTTP/1.1 404 Not Found]
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [opencast_event]","resource.type":"index_or_alias","resource.id":"opencast_event","index_uuid":"_na_","index":"opencast_event"}],"type":"index_not_found_exception","reason":"no such index [opencast_event]","resource.type":"index_or_alias","resource.id":"opencast_event","index_uuid":"_na_","index":"opencast_event"},"status":404}
		at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:326) ~[!/:?]
		at org.elasticsearch.client.RestClient.performRequest(RestClient.java:296) ~[!/:?]
		at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1632) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1602) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1572) ~[!/:?]
		at org.elasticsearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1088) ~[!/:?]
		at org.opencastproject.elasticsearch.impl.AbstractElasticsearchIndex.executeQuery(AbstractElasticsearchIndex.java:534) ~[!/:?]
		at org.opencastproject.elasticsearch.index.ElasticsearchIndex.getByQuery(ElasticsearchIndex.java:576) ~[!/:?]
		at org.opencastproject.adminui.endpoint.AbstractEventEndpoint.getEvents(AbstractEventEndpoint.java:2433) [!/:?]
		at jdk.internal.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[?:?]
		at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
		at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
```

To trigger this I started a fresh OC 11 build, processed a single recording, cleared the index, then rebuilt via the rest docs.

Edit: Also, the reindex *still appears to work* since the event reappeared in the admin page after the reindex ¯\\_(ツ)_/¯

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md]: You can't do _nothing_, exactly. If you want to keep using the Admin UI index without migrating it, you would have to change the index name in the config accordingly, as is described in Option 3. Then, the External API will also use the Admin UI index (which is why it's the last option, since this is a bit unclean).

If you don't do anything, Opencast would create an empty index under the new name for both, so then you basically have to do a rebuild.

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md]: Only the rebuild needs to happen during runtime, however, Option 1 clearly states this.
I put this before the restart since people should read the whole thing and decide what they want to do before starting Opencast. So I'd rather not link to the options directly. However, I'm gonna make the instructions a little more clear.

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md:28]: Added an 'if you haven't already done so' to make things clearer.

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md:43]: Done.

Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md]: Changed the order so the important notices come first. :)

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md]: Referenced the deletion script for the relevant commands. Made really sure people don't just run it...

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md]: Attempted to do this, but damn, is it hard to make this concise, but safe.

File: docs/guides/admin/docs/upgrade.md
Comment (KatrinIhler) [reply docs/guides/admin/docs/upgrade.md:74]: Added explanation.

**Comment by KatrinIhler** [comment]
@gregorydlogan I think I addressed all your review findings now. Writing upgrade docs for something like this is hard... but as long as people actually read the damn thing, they should be fine.
I could not reproduce that Exception of yours even though I did the exact same things in the same order. This almost looks like an Elasticsearch glitch? I mean, clearly _some_ event index was available before. This could only happen if the Elasticsearch was restarted and doesn't keep its state (my docker setup does this) or if you would change the index identifier in the middle of things, which can't happen without a module restart (or maybe even OC restart). Very odd... We're running this code at SWITCH since the summer and haven't seen something like this yet.

**Review by gregorydlogan — APPROVED** [review]

**Comment by mliradelc** [comment]
All issues are addressed, and is proven in production since summer.

Thanks for the awesome job 😄

Merged!

**Pull request merged**
